modules/storages/app/services/storages/file_link_sync_service.rb

```diff
@@ -0,0 +1,218 @@
+  # In case of completely deleted file or other errors we might also get:
+  # "24" = { "status" => "Not Found", "statuscode" => 404 }
+  def sync_single_file(file_link, origin_file_info_hash)
+    file_link.origin_mime_type = origin_file_info_hash["mimetype"]
+    file_link.origin_created_by_name = origin_file_info_hash["owner_name"]
+    file_link.origin_name = origin_file_info_hash["name"]
+    file_link.origin_created_at = Time.zone.at(origin_file_info_hash["ctime"])
+    file_link.origin_updated_at = Time.zone.at(origin_file_info_hash["mtime"])
+
+    file_link
+  end
+
+  private
+
+  # Sync a Nextcloud storage
+  def sync_nextcloud(storage_id, storage_file_links)
+    # Get the OAuthClientToken that will authenticate us against Nextcloud
+    oauth_client = OAuthClient.find_by(integration_id: storage_id)
+    connection_manager = ::OAuthClients::ConnectionManager.new(user: @user, oauth_client:)
+    connection_manager_service_result = connection_manager.get_access_token # No scope for Nextcloud...
+    unless connection_manager_service_result.success
+      set_error_for_file_links(storage_file_links)
+      return
+    end
+
+    @oauth_client_token = connection_manager_service_result.result
+
+    storage_origin_file_ids = storage_file_links.map(&:origin_id)
+    nextcloud_request_result = connection_manager.request_with_token_refresh do
+      response = request_files_info(storage_origin_file_ids)
+      # Parse HTTP response an return a ServiceResult with:
+      #   success: result= Hash with Nextcloud filesinfo (name of endpoint) data
+      #   failure: result= :error or :not_authorized
+      parse_files_info_response(response)
+    end
+    # Pass errors from Nextcloud into service result
+    @service_result.merge!(nextcloud_request_result)
+
+    if nextcloud_request_result.failure?
+      set_error_for_file_links(storage_file_links)
+      return
+    end
+
+    set_file_link_permissions(storage_file_links, nextcloud_request_result.result)
+  end
+
+  def set_error_for_file_links(storage_file_links)
+    @service_result.result += storage_file_links.each { |file_link| file_link.origin_permission = :error }
+    @service_result.success = false
+  end
+
+  # Check the Nextcloud status of a list of files:
+  # The endpoint returns "canonical" infos per file. Canonical means that the attributes are the same
+  # as the owner sees them and not the current user. If it was the current user the file name could be
+  # different for that user as a shared file can be renamed.
+  # @param file_ids An array of Nextcloud IDs of files to check
+  # @return A HTTP response or an Exception object
+  #
+  # curl -H "Accept: application/json" -H "Content-Type:application/json" -H "OCS-APIRequest: true"
+  # 		-u USER:PASSWD http://my.nc.org/ocs/v1.php/apps/integration_openproject/filesinfo
+  # 		-X POST -d '{"fileIds":[FILE_ID_1,FILE_ID_2,...]}'
+  def request_files_info(file_ids)
+    host = @oauth_client_token.oauth_client.integration.host
```

Comment:
I don't like the fact, that `@oauth_client_token` is exposed as a instance variable, just to be used here in the private method. Instead, you should do the following:

- in the method `sync_nextcloud` get the storage record.
- from the storage record get the `oauth_client` to be used in the connection manager and the `host` to be passed into this method `request_files_info`
- from the access token result get the access token and pass it into this method `request_files_info`
- avoid exposing this variable